app/models/box_form.rb

```diff
@@ -0,0 +1,95 @@
+class BoxForm
+  attr_reader :box, :batch_numbers
+
+  delegate :purpose, :purpose=, to: :box
+  delegate :model_name, :errors, to: :box
+
+  def self.build(navigation_context, params = nil)
+    box = Box.new(
+      institution: navigation_context.institution,
+      site: navigation_context.site,
+    )
+
+    if params
+      batch_numbers = params.delete(:batch_numbers) || {}
+      box.attributes = params
+    else
+      batch_numbers = {}
+    end
+
+    new(box, batch_numbers)
+  end
+
+  def initialize(box, batch_numbers)
+    @box = box
+    @batch_numbers = batch_numbers
+  end
+
+  def batches=(relation)
+    records = relation.to_a
+
+    @batches = Hash[@batch_numbers.map do |key, batch_number|
+      [key, records.find { |b| b.batch_number == batch_number }]
+    end]
```

Comment:
Looks like this could be simplified using `transform_values`:
```suggestion
    @batches = @batch_numbers.transform_values do |batch_number|
      records.find { |b| b.batch_number == batch_number }
    end
```